### ISSUE TITLE
BLIP2 VQA Questions

### ISSUE THREAD

**Issue by sachit-menon** [salesforce/LAVIS#99]
Hi, thanks for the great work on BLIP2, and also for open-sourcing the model and code!

I was trying to apply 'blip_t5' with model type "pretrain_flant5xxl" to VQA settings, and I suspect I'm missing something because so far I haven't been able to come close to the paper results -- in particular, I am getting 33.55 on GQA vs the paper's 44.7. 

I've tried with the following settings, per the paper: 
- Prompt: "Question: {} Short answer:"
- Beam search with width 5
- Length penalty -1

Otherwise all settings are left to the defaults. I'm using the `.generate` function per the Jupyter notebook. 

Are there other settings, postprocessing, evaluation details, etc that are important for performance, or specific to GQA?

I know the full VQA evaluation script is coming soon https://github.com/salesforce/LAVIS/issues/97#issuecomment-1416745600 , and I'm happy to just look off of that if it's released in the next few days. Otherwise it would be great to have some of these details while the full script is being prepared, if possible! Thanks again and great work!

**Comment by LiJunnan1992** [comment]
Can you reproduce the VQAv2 performance? Have you observed this gap for other models?

These are the same settings that I used. I don't think any hyperparameter can cause this much a gap.
- Prompt: "Question: {} Short answer:"
- Beam search with width 5
- Length penalty -1

**Comment by sachit-menon** [comment]
Thanks for your quick reply. I haven't tried VQAv2 or other models yet, but I can try them tomorrow and let you know. 

If it helps narrow down what could be the issue -- the results are all qualitatively reasonable, they're just (often) not correct. Here are some examples from the GQA train split. 

<img width="284" alt="image" src="https://user-images.githubusercontent.com/21482705/216802687-c11d75f6-1aaf-429e-83fa-b9a33af2c4f6.png">

**Comment by LiJunnan1992** [comment]
This #100 might be the reason for generating wrong answers?

**Comment by sachit-menon** [comment]
It's definitely possible, that's why I wanted to check about that too. But, because I wondered about that, I did a subset of 2000 from GQA one by one with the original unmodified code and the performance between the two matched, plus it seemed correct to you in the other thread; so I'd be surprised if it were the cause, but not impossible. Running the whole dataset one by one is too slow -- any other recommended way to batch things? 

(Also, VQAv2 should finish tonight but the results actually look reasonable so far (halfway through val).)

**Comment by sachit-menon** [comment]
Hi Junnan, I'm really sorry for the confusion. I realized something I was doing resulted in the `length_penalty` not being passed through correctly. With the correct `length_penalty=-1.`, I get 42.2 on GQA testdev, which is much closer to the paper result.

Now I'm pretty confident everything is correct (only change is the batching from #100) but the performance is still 2.5% less than the lowest Flan model in the paper, which is non-negligible for GQA -- any ideas on what might still be missing? (What was your seed, just to standardize?)

Sorry again and thanks!

**Comment by yonatanbitton** [comment]
Thanks for the clarification @sachit-menon .
It's also related to the issue I opened: [https://github.com/salesforce/LAVIS/issues/97](url)
I think it will be very helpful to clarify which hyperparameters should be called for every task. For example that length penalty should be -1, while it isn't -1 in the [default parameters](https://github.com/salesforce/LAVIS/blob/main/lavis/models/blip2_models/blip2_t5.py#L146). Without this clarification people can report incorrect accuracies using BLIP2 that will cause confusion, and the reported results will be below the actual results that can be reported using the correct parameters. @LiJunnan1992 is something like that possible? 🙂 
Thanks

**Comment by LiJunnan1992** [comment]
We are working on BLIP-2 VQA evaluation scripts @yonatanbitton. Meanwhile please refer to our paper for hyperparameters.

**Comment by kebijuelun** [comment]
> We are working on BLIP-2 VQA evaluation scripts @yonatanbitton. Meanwhile please refer to our paper for hyperparameters.

Hi, thanks for the wonderful work. I have not found  the VQAv2 evaluation code in https://github.com/salesforce/LAVIS/tree/main/run_scripts/blip2/eval . When will the VQAv2 evaluation code (for blip2) be open source？

**Comment by dxli94** [comment]
@kebijuelun we are working on the evaluation script. Please keep an eye on https://github.com/salesforce/LAVIS/tree/dev0.4.

Expecting to finish in the coming few days. Thanks.

**Comment by sachit-menon** [comment]
Hi @dxli94 , I saw that some of this work (VQAv2 and OKVQA) has landed now -- thanks for that! 

I'm particularly interested in GQA, and still unable to reproduce that result (42.2% vs 44.7%, which would no longer be SOTA as it is a bit less than your own group's work on PNP-VQA).

Any chance that GQA can also be added to this soon? I would really appreciate if it were made a priority and it would finish off the VQA set of experiments :) .

**Comment by LiJunnan1992** [comment]
@sachit-menon GQA has been added.

**Comment by sachit-menon** [comment]
Amazing, thanks for the quick update! I'll look into this today. (From a glance at the eval yaml, one difference might be from `max_len=10` in the script vs the default value of 30, but I'll find out properly.)

**Comment by sachit-menon** [comment]
With your script I am getting about the accuracy reported in the eval yaml (~44.) for GQA, thanks again!

What is the correct way to run on test-dev? (I noticed there's also a split "test" instead of "val" which I imagined was the `test-dev` split, so I tried to run that but got an error in dataloading that it just produced Nones.) I am particularly interested in getting the 44.7 that the XXL model achieves.
